8241 ipv6 client ip

markrwilliams opened the pull request:
## Contributor Checklist:

* [x] There is an associated ticket in Trac. http://twistedmatrix.com/trac/ticket/8241
* [x] The changes pass minimal style checks (see: https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted )
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles )
* [x] I have updated the automated tests.

markrwilliams committed:
Failing test demonstrating that an IPv6 address is returned as None.

markrwilliams committed:
t.w.http.Request.getClientIP now returns the host for IPv6Addresses.

markrwilliams committed:
Newfragment.

markrwilliams commented:
Also http://twistedmatrix.com/trac/ticket/7704

glyph commented:
(I think) I've disabled CircleCI status updates because the configuration is completely broken.  @adiroiban If you'd like to turn it back on please fix it so it doesn't fail every build :).

glyph committed:
Merge branch 'trunk' into 8241-ipv6-client-ip

markrwilliams committed:
Remove extra space.

glyph committed:
Merge branch 'trunk' into 8241-ipv6-client-ip

adiroiban commented on src/twisted/web/http.py:
```diff
@@ -1386,7 +1386,7 @@ def getClientIP(self):
         @rtype: C{str}
         """
-        if isinstance(self.client, address.IPv4Address):
+        if isinstance(self.client, (address.IPv4Address, address.IPv6Address)):
```
This is fine... but maybe we should use  `_IPAddress` here... I don't know which is better.

markrwilliams replied:
I'm reluctant to use a private implementation even though this is inside Twisted.

adiroiban commented on src/twisted/web/http.py:
```diff
@@ -1386,7 +1386,7 @@ def getClientIP(self):
+        if isinstance(self.client, (address.IPv4Address, address.IPv6Address)):
             return self.client.host
         else:
             return None
```
I know that this is existing code, but when is expected to return `None`.

The docstring is not clear.

Can this ever return `None`? In which case? Is there a test for this?

The code coverate, reports that this branch is executed  but I could not found the explicit tests https://codecov.io/gh/twisted/twisted/src/89e71172c2e4b2f923d3f3fa49922d50262394df/src/twisted/web/http.py#L1392

I saw that it is called in twisted.web.test.test_web.test_etagMatched due to a log which is trigerred.

Can we have a test describing why we have this execution branch?

markrwilliams replied:
> Can this ever return None?

It absolutely can!  Try this:

```twist web --port=unix:/tmp/sock --wsgi wsgiref.simple_server.demo_app```

Connect to it like so:

```curl --unix-socket /tmp/sock http://example.invalid/```

Now you'll get a traceback from ```twisted.web```:

```
...
          File "/home/mrw/src/python/twisted/src/twisted/web/wsgi.py", line 303, in __init__
            'REMOTE_ADDR': _wsgiString(request.getClientIP()),
          File "/home/mrw/src/python/twisted/src/twisted/web/wsgi.py", line 85, in _wsgiString
            return string.decode("iso-8859-1")
        builtins.AttributeError: 'NoneType' object has no attribute 'decode'
```

This is mentioned in [a comment on 7704](https://twistedmatrix.com/trac/ticket/7704#comment:2), which was linked to from [this ticket](https://twistedmatrix.com/trac/ticket/8241) and also captured in [5406](https://twistedmatrix.com/trac/ticket/5406).

adiroiban replied:
Great work :) Feel free to merge this as it is. Having a test for this is optional and I hope it will be done in the follow-up ticket

adiroiban reviewed (approved):
Thanks for working on this.

It looks much better than trunk :)

Is there a test for the non IPv6 or IPv4 execution branch?

Can we have HTTP requests over non-TCP/IP stack?

I don't know why the OSX failed on installing tox.
I saw that OSX pass after this on a separate branch.
See if they still fail after merging with trunk.

Thanks!

markrwilliams committed:
Merge branch 'trunk' into 8241-ipv6-client-ip

markrwilliams committed:
Test that getClientIP returns None for non-TCP connections.

markrwilliams committed:
This fixes two issues.  Test 8241, and add a newsfragment for 7704.

The original tests were for 7704 and the included 8241 newsfragment
incorrectly described them.  Now there are two newsfragments and a
test for 8241.

markrwilliams commented:
If this is merged, its `Fixes` should include two tickets: https://twistedmatrix.com/trac/ticket/7704 and https://twistedmatrix.com/trac/ticket/8241

markrwilliams committed:
Merge branch 'trunk' into 8241-ipv6-client-ip

markrwilliams committed:
Merge branch 'trunk' into 8241-ipv6-client-ip

markrwilliams committed:
Merge branch 'trunk' into 8241-ipv6-client-ip

markrwilliams commented:
@rodrigc This was in the review queue - be sure to remove it to get review credit.

markrwilliams commented:
Also, per https://github.com/twisted/twisted/pull/975#issuecomment-375925298, this needs to close 7704 as well.
